### PR TITLE
Avoid over-linking #1293

### PR DESCRIPTION
**Pull request by nedbat**
Summarizing and codifying the advice discussed in https://github.com/python/docs-community/issues/52

----
📚 Documentation preview 📚: https://cpython-devguide--1294.org.readthedocs.build/

### CODE REVIEW

File: documentation/style-guide.rst [hunk 1]
```diff
@@ -124,11 +124,37 @@ explanation.
...
+=====
+
+Linking words to more information about those words is a powerful tool for
+helping people navigate documentation, but links can be over-used.
+
+Links should be provided only if they help the user.  If a link will not help
+the user, do not create a link.  You may need to suppress a link that Sphinx
+would have created automatically.
+
+Generally, a link should be provided for the first use of a term in a unit This
```
Comment (willingc) [inline documentation/style-guide.rst:142]:
```suggestion
Generally, a link should be provided for the first use of a term in a unit, such as a section or paragraph. This
```

File: documentation/style-guide.rst [hunk 2]
```diff
@@ -124,11 +124,37 @@ explanation.
...
+is not a hard and fast rule.  Sometimes the second mention is a more
+appropriate jumping off point.  Some units are long enough to have a few
```
Comment (willingc) [inline documentation/style-guide.rst:144]:
```suggestion
appropriate for a link.  Some units are long enough to have a few
```

Comment (willingc) [reply documentation/style-guide.rst:144]: @nedbat I'm not sure that "jumping off point" is commonly used globally.

File: documentation/style-guide.rst [hunk 3]
```diff
@@ -124,11 +124,37 @@ explanation.
...
 Please consult the `Diátaxis <https://diataxis.fr/>`_ guide for more
 detail.
 
+Links
+=====
+
+Linking words to more information about those words is a powerful tool for
+helping people navigate documentation, but links can be over-used.
+
+Links should be provided only if they help the user.  If a link will not help
```
Comment (willingc) [inline documentation/style-guide.rst:138]:
```suggestion
Links should be provided only if they help the reader.  If a link will not help
```

File: documentation/style-guide.rst [hunk 4]
```diff
@@ -124,11 +124,37 @@ explanation.
...
+the user, do not create a link.  You may need to suppress a link that Sphinx
```
Comment (willingc) [inline documentation/style-guide.rst:139]: I'm unclear on suppressing a link. Perhaps omit from this paragraph and move further down the section.

**Review by willingc — COMMENTED** [review]
Thanks Ned!

**Review by jablonskidev — APPROVED** [review]

File: documentation/style-guide.rst (hunk 2 above)
Comment (nedbat) [reply documentation/style-guide.rst:144]: Done.

File: documentation/style-guide.rst (hunk 4 above)
Comment (nedbat) [reply documentation/style-guide.rst:139]: Done.

**Comment by nedbat** [comment]
Changes made, and I added another rule: no links in section titles.

**Review by hugovk — APPROVED** [review]

File: documentation/markup.rst [hunk 5]
```diff
@@ -30,11 +30,21 @@ external links          ```Link text <https://example.com>`_``      :ref:`hyperl
...
 issues                  ``:gh:`ID```, ``:issue:`ID```               :ref:`roles`
 CPython source          ``:source:`PATH```                          :ref:`roles`
 comments                ``.. a comment``                            :ref:`comments`
 ======================= =========================================== ====================
 
+Notes:
+
+(1)
+   For an only-last-part reference with a suppressed link,
+   ``:role:`~!hidden.visible``` makes more semantic sense, but it causes
```
Comment (merwok) [inline documentation/markup.rst:43]: semantic (adj): related to meaning, or to the study of meaning

Therefore semantic sense or semantic meaning is redundant 🙂 

```suggestion
   ``:role:`~!hidden.visible``` makes more sense, but it causes
```

File: documentation/markup.rst [hunk 6]
```diff
@@ -30,11 +30,21 @@ external links          ```Link text <https://example.com>`_``      :ref:`hyperl
...
+   a warning as Sphinx tries to look up the reference ``!hidden.visible``
+   which does not exist.  The shorter form ``:role:`!visible`` renders as
+   desired and will build successfully.
```
Comment (ezio-melotti) [inline documentation/markup.rst:46]: I'm not sure this should be included here.  The quick reference lists commonly used markup, and what's being described here is not very common.

I would move this paragraph in the linked ["roles" section](https://devguide.python.org/documentation/markup/#roles), where the nuances of using `!` and `~` are covered in more detail.

File: documentation/style-guide.rst [hunk 7]
```diff
@@ -124,11 +124,39 @@ explanation.
...
-  :ref:`python:unicode-howto`
+  :ref:`python:unicode-howto`.
 
 Please consult the `Diátaxis <https://diataxis.fr/>`_ guide for more
 detail.
 
+Links
+=====
+
+Linking words to more information about those words is a powerful tool for
```
Comment (ezio-melotti) [inline documentation/style-guide.rst:135]: I would rephrase this sentence.

File: documentation/style-guide.rst [hunk 8]
```diff
@@ -124,11 +124,39 @@ explanation.
...
+
+Do not use links in section headers.  They distract from the title of the
+section.  The term will be mentioned in the paragraph text and can be linked
+from there.
+
+Sphinx provides ways to automatically add links to references, and a way to
+suppress the link.  Using roles like ``:func:`map``` will link to the
+documentation for ``map``.  You can suppress the link while keeping the
+semantic presentation of the function name by adding an exclamation point
+prefix: ``:func:`!map```.
```
Comment (ezio-melotti) [inline documentation/style-guide.rst:158]: Here you could link to the "roles" section of `markup.rst`.

File: documentation/markup.rst (hunk 6 above)
Comment (nedbat) [reply documentation/markup.rst:46]: It took a few developers a while to work out what happened when trying to combine `~!`, so I think it will be helpful to leave here.

File: documentation/style-guide.rst (hunk 7 above)
Comment (nedbat) [reply documentation/style-guide.rst]: Done

File: documentation/style-guide.rst (hunk 8 above)
Comment (nedbat) [reply documentation/style-guide.rst]: Done

File: documentation/markup.rst (hunk 6 above)
Comment (ezio-melotti) [reply documentation/markup.rst:46]: What I'm thinking is:
* the quick reference table is meant mostly for readers that already know about the existence and functioning of those constructs, but don't quite remember the syntax
* all further details and explanations shouldn't (and aren't) covered here, but in the sections linked in the third column
* if users look at the "roles" section (where `~` and `!` are explained) for ways to combine `~`/`!` they won't find it there, unless you duplicate the note there
* the wording of `roles w/o link, short` is not too clear, since `short` doesn't mirror `only last part` (understandably, due to lack of space)
* a reader might be wondering how to combine `~` and `!` together, but `` :role:`!visible` `` doesn't seem an obvious answer to the question (unless they keep scrolling and read the note), especially since it doesn't have any apparent difference from  `` :role:`!target` ``
* the consequences of combining `~` and `!` are a Sphinx implementation detail that shouldn't matter to the reader, so I would omit them from the note

My suggestion is to add:
```rst
* Combining ``~`` and ``!`` (e.g. ``:meth:`~!Queue.Queue.get```) is not supported.
  You can obtain the same result by simply using `!` and the last component of
  the target (e.g. ``:meth:`!get```).
```
in the "roles" section, after the bullet points that already explain `~` and `!` and use `Queue.Queue.get` as an example.

The entry in the table could be removed altogether, since it's not a very common construct (trusting that the reader will follow the link looking for more information).  If you prefer to keep it, the table should be updated by:
* widening the first column and using `roles w/o link, only last part` to mirror the text of `roles w/ only last part`
* reordering the entries so that `roles w/o link` comes first, `roles w/ only last part` next, and `roles w/o link, only last part` just after, so that the use of `!visible` is more evident.

Comment (willingc) [reply documentation/markup.rst:46]: @nedbat Let me take some time to read @ezio-melotti's comment. Ezio makes some good points especially considering the original intent of the chart. There needs to be a new doc contributor cheat sheet (that is useful to current contributors) to guide folks on many of these markup constructs.

Comment (nedbat) [reply documentation/markup.rst]: @ezio-melotti I like your suggestion, and have made it in the latest commit.

**Review by ezio-melotti — APPROVED** [review]

File: documentation/markup.rst [hunk 9]
```diff
@@ -753,6 +753,10 @@ versatile:
   In HTML output, the link's ``title`` attribute (that is e.g. shown as a
   tool-tip on mouse-hover) will always be the full target name.
 
+* Combining ``~`` and ``!`` (for example, ``:meth:`~!Queue.Queue.get```) is not
+  supported.  You can obtain the same result by simply using ``!`` and the last
```
Comment (hugovk) [inline documentation/markup.rst:757]: Don't say simply: 
* https://www.youtube.com/watch?v=gsT2BBWBVmM
* https://blog.knowledgeowl.com/blog/posts/dont-say-simply-jim-fisher/
* https://jameshfisher.com/2017/02/22/dont-use-simply/


```suggestion
  supported.  You can obtain the same result by using ``!`` and the last
```

File: documentation/style-guide.rst [hunk 10]
```diff
@@ -124,11 +124,39 @@ explanation.
...
+
+Generally, a link should be provided for the first use of a term in a unit,
+such as a section or paragraph. This is not a hard and fast rule.  Sometimes
+the second mention is more appropriate for a link.  Some units are long enough
+to have a few repeated links.  Use judgement to decide when a link will help
+the reader.
+
+Do not use a link when the link would point to the current unit.  It's natural
+to use the name of a function in the documentation for the function, but a link
+on that function name that simply reloads the section the user is already
```
Comment (hugovk) [inline documentation/style-guide.rst:147]: (I think this "simply" is fine, it's describing the redundant behaviour.)

**Comment by hugovk** [comment]
Thanks!

**Comment by willingc** [comment]
Thanks @nedbat
